Pkg table

pypingou opened the pull request:
These branch introduces a `packages` table, linked to the `files` table and which only stores the package name.

This should allow @charulagrl to list all the packages much more easily and quickly and thus optimize the output of the /packages/ endpoint on shumgrepper.

pypingou committed:
Create the `packages` table storing the package name and link them to the `files` table

pypingou committed:
Add SQL script create the link between the packages and the files tables

puiterwijk commented on summershum/model.py:
```diff
@@ -55,7 +64,11 @@ class File(BASE):
-    pkg_name = sa.Column(sa.Text, index=True, nullable=False)
+    pkg_name = sa.Column(
+        sa.Text,
+        sa.ForeignKey('Package.pkg_name', onupdate='CASCADE'),
```
This should point to packages.pkg_name, given that that is the tablename of the Package class.

pypingou committed:
Fix the foreign key reference to point to the table rather than the object

Thanks @puiterwijk for finding it out

pypingou committed:
Add an intermediary table between Package and File which lists the Releases

With this scheme we are going from
  Package -> File
to
  Package -> Release -> File

Which is a good enhancement from a DB scheme pov

pypingou committed:
Fix keyword primary_key is not primary

pypingou committed:
Add missing import on sqlalchemy.orm for relation

pypingou committed:
Adjust the SQL script to update the current model to the new schema

ralphbean commented:
Any thoughts on using alembic instead of sql scripts to be run by hand?

I dream of the day we can reference an alembic ansible module, something like:

``` YAML
alembic:  state=latest config=/usr/share/summershum/alembic.ini
```

pypingou commented:
I'd be +1 to use an alembic script it will just require a little more work

ralphbean commented:
Cool - I won't block it if we don't want to do the alembic work now.  But it seems like we're standardizing in that direction on our other projects.

ralphbean commented on 0.1.6_upgrade.sql:
```diff
@@ -0,0 +1,13 @@
+INSERT INTO packages (pkg_name) VALUES (SELECT DISTINCT pkg_name FROM files);
```
I get a syntax error in sqlite for this line.

ralphbean commented on summershum/model.py:
```diff
@@ -55,18 +90,12 @@ class File(BASE):
-
     created_on = sa.Column(sa.DateTime, default=datetime.datetime.utcnow)
 
+    tarball = relation('Release', remote_side=[tarball], backref='files')
```
The `remote_side` here doesn't work -- raises an Exception.

ralphbean commented on 0.1.6_upgrade.sql:
```diff
@@ -0,0 +1,13 @@
+INSERT INTO packages (pkg_name) VALUES (SELECT DISTINCT pkg_name FROM files);
+
+INSERT INTO releases (pkg_name, tarball, tar_sum) VALUES (
+    SELECT DISTINCT pkg_name, tar_file, tar_sum FROM files
```
I don't think there is a `tar_file` column here to select.  It seems like it was renamed to `tarball` in 165ff5e04

ralphbean committed:
Typofix in table name.

ralphbean committed:
Fix foreign relation between file and release.

ralphbean committed:
Remove some pkgdb2 stuff that copy paste in here somehow.

ralphbean commented:
OK.  I added alembic stuff and fixed up some other stuff.  :+1: from me.

@pypingou do you want to have another look at it and then either merge it or comment?

pypingou commented:
@ralphbean forgot to push the alembic commits?

ralphbean committed:
Add forgotten alembic stuff\!\!\!

ralphbean commented:
Ah, so I did.  Just pushed them in 1637f52

pypingou commented:
Looks good to me :)

ralphbean commented:
:fireworks: thanks! :fireworks:
